### PR TITLE
added postprocessing output for truss

### PR DESCRIPTION
**Pull request by KlausBSautter**
see #1133

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.hpp [hunk 1]
```diff
@@ -152,6 +162,8 @@ namespace Kratos
 		bounded_vector<double,msLocalSize> CalculateBodyForces();  
 
 		bool ReturnIfIsCable();
+		
+		void AddPrestressLinear(VectorType& RHS);
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.hpp]: Not consistent with the Definition in the cpp

File: applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp [hunk 2]
```diff
@@ -408,6 +410,75 @@ namespace Kratos
...
+		std::vector< array_1d<double, 3 > >& rOutput,
+		const ProcessInfo& rCurrentProcessInfo) 
+		{
+		if (rVariable == FORCE)
+				{
+					Vector truss_forces = ZeroVector(msDimension);
+					truss_forces[2] = 0.00;
+					truss_forces[1] = 0.00;
+
+					const double internal_strain_gl = this->CalculateGreenLagrangeStrain();
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]: It seems like this is only needed for the nonlin case
Can you make an if else?
The if for the linear case is already there

File: applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp [hunk 3]
```diff
@@ -408,6 +410,75 @@ namespace Kratos
...
+
+	void TrussElement3D2N::AddPrestressLinear(VectorType& rRightHandSideVector)
+	{
+		KRATOS_TRY;
+		bounded_matrix<double,msLocalSize,msLocalSize>
+		 transformation_matrix = ZeroMatrix(msLocalSize, msLocalSize);
+		this->CreateTransformationMatrix(transformation_matrix);
+		double S_pre = 0.00;
+		if (this->GetProperties().Has(TRUSS_PRESTRESS_PK2)) {
+			S_pre = this->GetProperties()[TRUSS_PRESTRESS_PK2];
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]: Name is not really clear, can you use sth more descriptive?

**Review by philbucher — CHANGES_REQUESTED** [review]

**Comment by philbucher** [comment]
and a small test is missing
I suggest to integrate it in the existing tests

**Comment by pooyan-dadvand** [comment]
Just to say that I agree with comments of @philbucher

File: applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp [hunk 4]
```diff
@@ -408,6 +410,75 @@ namespace Kratos
...
+					const double l = this->CalculateCurrentLength();
+					const double E = this->GetProperties()[YOUNG_MODULUS];
+					const double A = this->GetProperties()[CROSS_AREA];
+
+					double s_pre = 0.00;
+					if (this->GetProperties().Has(TRUSS_PRESTRESS_PK2)) {
+						s_pre = this->GetProperties()[TRUSS_PRESTRESS_PK2];
+					}
+
+					truss_forces[0] = ((E*internal_strain_gl + s_pre) * l * A) / L0;
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]: i see that you are using a linear displacement measure, l/l0. what kind of formulation are u using here?

also why do you need to do something different in the linear case?

Comment (KlausBSautter) [reply applications/StructuralMechanicsApplication/custom_elements/truss_element_3D2N.cpp]: for linear strain meassurement only the change of length as a projection wrt the reference configuration is taken. For non-linear cases I need the actual change of length

**Comment by KlausBSautter** [comment]
I try to call CalculateOnIntegrationPoints(const Variable<array_1d<double, 3 > >& rVariable, std::vector< array_1d<double, 3 > >& rOutput, const ProcessInfo& rCurrentProcessInfo) in the test_patch_test_truss.py file. How do I initialize rOutput in Pyhton so that the function can work with it?

**Comment by KlausBSautter** [comment]
in test_quadratic_element.py it is written like this:
strains = elem.CalculateOnIntegrationPoints(KratosMultiphysics.GREEN_LAGRANGE_STRAIN_VECTOR, mp.ProcessInfo)
which is strange because this function need another argument 
:(

**Comment by KlausBSautter** [comment]
+ is a void funtion

**Comment by KlausBSautter** [comment]
@philbucher I changed the lines you referred to and added tests

File: applications/StructuralMechanicsApplication/tests/test_patch_test_truss.py [hunk 5]
```diff
@@ -222,6 +223,14 @@ def _check_results_nonlinear(self,mp,timestep,Force_i):
         P_i = ((EA/(2*L3))*(disp_temp*disp_temp +2*1*disp_temp)*(disp_temp+1))
         self.assertAlmostEqual(P_i,Force_i,1)
 
+    def _check_pre_stress_output(self,mp,force,tollerance=9):
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_patch_test_truss.py]: typo

**Review by philbucher — APPROVED** [review]
minor

**Pull request merged**